util/HalideTraceDump.cpp

```diff
@@ -0,0 +1,403 @@
+#include "HalideTraceUtils.h"
+#include "HalideBuffer.h"
+#include "halide_image_io.h"
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdint.h>
+#include <vector>
+#include <map>
+#include <assert.h>
+#include <string>
+#include <unistd.h>
+#include <fcntl.h>
+#include <string.h>
+
+/** \file
+ *
+ * A tool which can read a binary Halide trace file from stdin, and dump files containing the final pixel values
+ * recorded for each traced Func.
+ *
+ * Currently dumps into supported Halide image formats.
+ */
+
+using std::map;
+using std::vector;
+using std::string;
+using Halide::Runtime::Buffer;
+
+struct BufferOutputOpts {
+    enum OutputType {
+        PNG = 0,
+        JPG,
+        PGM
+    };
+
+    enum OutputType type;
+};
+
+int halide_copy_to_host(void *user_context, struct halide_buffer_t *buf) {
+    return 0;
+}
+
+struct Value {
+    uint8_t payload[8];
+    uint8_t defined[8];
+
+    Value() { defined[0] = 0; }
+    Value(Packet *p, int lane_idx=0) {
+        defined[0] = 1;
+        int bytes = p->type.bytes();
+        unsigned long addr = (unsigned long) p->value();
+        addr += (bytes*lane_idx);
+        memcpy((void*)payload, (void*)addr, bytes);
+    }
+
+    template<typename T>
+    T get_as(halide_type_t type) {
+        return value_as<T>(type, (void*)payload);
+    }
+};
+
+struct FuncInfo {
+    int min_coords[16];
+    int max_coords[16];
+    int extents[16];
+    int dimensions;
+    halide_type_t type;
+    Value * values;
```

Comment:
Are you suggesting to make a typedef alias somewhere to the Value type called Values, to clarify that it is plural in this usage?